sunxspex/thermal.py

```diff
@@ -0,0 +1,699 @@
+    The temperature of the plasma.
+    Can be scalar or 1D of any length. If not scalar, the flux for each temperature
+    will be calculated. The first dimension of the output flux will correspond
+    to temperature.
+
+emission_measure: `astropy.units.Quantity`
+    The emission measure of the plasma at each temperature.
+    Must be same length as temperature or scalar.
+
+abundance_type: `str` (optional)
+    Abundance type to use.  Options are:
+        1. cosmic
+        2. sun_coronal - default abundance
+        3. sun_coronal_ext
+        4. sun_hybrid
+        5. sun_hybrid_ext
+        6. sun_photospheric
+        7. mewe_cosmic
+        8. mewe_solar
+    The values for each abundance type is stored in the global
+    variable DEFAULT_ABUNDANCES which is generated by `setup_default_abundances`
+    function. To load different default values for each abundance type,
+    see the docstring of that function.
+
+relative_abundances: `tuple` of `tuples` of (`int`, `float`) (optional)
+    The relative abundances of different elements as a fraction of their
+    default abundances defined by abundance_type.
+    Each tuple represents the values for a given element.
+    The first entry represents the atomic number of the element.  If this is not
+    entered as an int, it is rounded to the nearest int.
+    The second entry represents the axis represents the fraction by which the
+    element's default abundance should be scaled.
+
+observer_distance: `astropy.units.Quantity` (Optional)
+    The distance between the source and the observer. Scales output to observer distance
+    and unit by 1/length.
+    Default=1 AU.
+
+Returns
+-------
+flux: `astropy.units.Quantity`
+    The photon flux as a function of temperature and energy.
+"""
+
+def setup_continuum_parameters(filename=None):
+    """
+    Define continuum intensities as a function of temperature.
+
+    Intensities are set as global variables and used in
+    calculation of spectra by other functions in this module. They are in
+    units of per volume emission measure at source, i.e. they must be
+    divided by 4 * pi R**2 to be converted to physical values where
+    R**2 is observer distance.
+
+    Intensities are derived from output from the CHIANTI atomic physics database.
+    The default CHIANTI data used here are collected from
+    `https://hesperia.gsfc.nasa.gov/ssw/packages/xray/dbase/chianti/chianti_cont_1_250_v71.sav`.
+    This includes contributions from thermal bremsstrahlung and two-photon interactions.
+    To use a different file, provide the URL/file location via the filename kwarg,
+    e.g. to include only thermal bremsstrahlung, set the filename kwarg to
+    'https://hesperia.gsfc.nasa.gov/ssw/packages/xray/dbase/chianti/chianti_cont_1_250_v70_no2photon.sav'
+
+    Parameters
+    ----------
+    filename: `str` (optional)
+        URL or file location of the CHIANTI IDL save file to be used.
+    """
+    global _CONTINUUM_GRID
```

Comment:
Wondering about the globals, what's the reasoning behind this?

Reply:
This is mainly for speed.  All parameters from the CHIANTI files are read & derived on import and placed in global variables.  This means the functions mostly only perform calculations that can't be done without the user-input parameters.  If this function is being called hundreds of times as part of an iterative fitting process, this speed gain will be important.  This was something Richard always emphasized when talking to us about sunxspex.

However, there is nothing stopping us having an `intensity_grid` keyword and having these global grids be the defaults.  That could make it easier for users to manipulate their input.  Also, regarding the continuum function, which mechanisms are included depends on the CHIANTI file.  The function itself is the same.  An `intensity_grid` kwarg could therefore make it easier for users to change the continuum emission mechanisms they want to include by using a grid from a different file.

All that being said, the use of global variables to construct the defaults and save computational time is still valid.

Reply:
Would a module level variables not work, just curious it really doesn't matter anyway?
`_CONTINUUM_GRID = setup_continuum_parameters()`

Reply:
It would for the default case where users aren't manually selecting the file they want to use.  If they doing that though, they would have to know to do
```python
from sunxspex import thermal
thermal._CONTINUUM_GRID = thermal.setup_continuum_parameters(myfile)
```
If they set the output to any other variable the data from `myfile` wouldn't be used which could be confusing.  But this is an API choice based on how we think the users will most likely want to use the software.  Either approach is technically valid.